src/CachePlugin.php

```diff
@@ -68,15 +68,42 @@ public function handleRequest(RequestInterface $request, callable $next, callabl
         $cacheItem = $this->pool->getItem($key);
 
         if ($cacheItem->isHit()) {
-            // return cached response
             $data = $cacheItem->get();
-            $response = $data['response'];
-            $response = $response->withBody($this->streamFactory->createStream($data['body']));
+            if (isset($data['expiresAt']) && time() > $data['expiresAt']) {
+                // This item is still valid according to previous cache headers
+                return new FulfilledPromise($this->createResponseFromCacheItem($cacheItem));
+            }
+
+            // Add headers to ask the server if this cache is still valid
+            if ($modifiedAt = $this->getModifiedAt($cacheItem)) {
+                $modifiedAt = new \DateTime('@'.$modifiedAt);
+                $modifiedAt->setTimezone(new \DateTimeZone('GMT'));
+                $request = $request->withHeader(
+                    'If-Modified-Since',
+                    sprintf('%s GMT', $modifiedAt->format('l, d-M-y H:i:s'))
+                );
+            }
 
-            return new FulfilledPromise($response);
+            if ($etag = $this->getETag($cacheItem)) {
+                $request = $request->withHeader(
+                    'If-None-Match',
+                    $etag
+                );
+            }
         }
 
+
         return $next($request)->then(function (ResponseInterface $response) use ($cacheItem) {
+            if (304 === $response->getStatusCode()) {
+                // The cached response we have is still valid
+                $data = $cacheItem->get();
+                $data['expiresAt'] = time() + $this->getMaxAge($response);
+                $cacheItem->set($data)->expiresAfter($this->config['cache_lifetime'] + $data['expiresAt']);
+                $this->pool->save($cacheItem);
+
+                return $this->createResponseFromCacheItem($cacheItem);
```

Comment:
We do now that it is a cache. Or else the `If-Modified-Since` and `If-None-Match` would not have been added. And correct me if I'm wrong but I don't think the server will answer 304 without those headers in the request.